SalesforceCommerceCloud/pwa-kit
# Handle validation error from commerce-sdk-isomorphic

UPDATE:

This PR no longer tries to configure react-query, since it's now configurable from the user land (see PR #734 ). 

Todos:
- [x] Remove example code

**How to test the changes:**

Assuming the default react-query options of `retry: false` set in the template/project:

- Validation error on http://localhost:3000/products will result immediately in the message "Something is wrong".
- Response error on http://localhost:3000/products/25502228Mxxx will also immediately print an error message.

---
There are 2 goals for this PR.. updating our hooks library such that:
- If there's an error, reuse the error thrown from `commerce-sdk-isomorphic`
- Stop react-query's retries if we're getting _validation_ errors ← most interesting goal

For the 2nd goal above, we can pass in a [`retry` function](https://tanstack.com/query/v4/docs/guides/query-retries) to react-query, which we'll set it up to work like this:
- If we get a `ResponseError` from `commerce-sdk-isomorphic`, we'll want to continue the retries.
- Otherwise, we'll stop the retries.

Ticket: [W-11677734](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE000015NKpHYAW/view)

## So which errors are considered to be validation errors?

I'd like to propose that the validation errors to be any error that is _not_ `ResponseError`. 
- It makes sense to think of 2 kinds of errors (response errors or not)
- As a side benefit, we don't need to make a new error class in the isomorphic sdk.

Let's consider the point of view from **commerce-sdk-isomorphic**. Why does it have `ResponseError` and the rest of errors thrown are just `Error`? (See [source code](https://github.com/SalesforceCommerceCloud/commerce-sdk-isomorphic/search?q=%22throw+new%22))
- ResponseError are something out of your control (something on the backend is down) and cannot be detected ahead of time
- while the rest of the errors are something that can be avoided and detected. For example: user error like calling a function with incorrect arguments.

Now, when you consider the point of view from **react-query**: why does it do retries? I believe they're for handling unexpected errors (like the ResponseError above).

So I think that we shouldn't worry about identifying other kinds of errors (validation errors, etc). Checking for `ResponseError` is already sufficient.

# How to Test-Drive This PR

First, `npm start` in the test-commerce project.

Then go to these urls and verify how we handle the errors. Check the browser console for details.
- http://localhost:3000/products
    - The page passes in wrong parameters to a function, so you should expect react-query to _not_ do any retry.
- http://localhost:3000/products/25502228Mxyz
    - This is not a good example, but for now assume that the product id is actually valid. But then the response from the API is 404. ~~So you should expect react-query to retry the request a few times.~~ The fetch retries now only happen for http 5xx errors, as discussed in this [comment](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/718#discussion_r974471781).

# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
